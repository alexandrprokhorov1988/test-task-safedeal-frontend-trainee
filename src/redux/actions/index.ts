import {setIsLoading, setOriginSizeImage, setCards, setComments} from './photoGridActions';
import {setIsLoadingComment} from './popupActions';
import {setIsOpenPopup} from './mainActions';

export default {
  setIsLoading,
  setOriginSizeImage,
  setCards,
  setComments,
  setIsLoadingComment,
  setIsOpenPopup
}
