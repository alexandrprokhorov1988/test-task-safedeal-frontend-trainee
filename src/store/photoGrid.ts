import {makeAutoObservable} from "mobx";
import {autoSave} from "../utils/helpers";

class PhotoGrid {
  public cards: any[] = [];

  public isLoading = false;

  public originSizeImage: { id?: number, url?: string } = {};

  public comments: { id: number, text: string, date: string }[] = [];

  constructor() {
    makeAutoObservable(this);
    autoSave(this, 'photoGridStore')
  }

  public setCards(newCard: any[]) {
    this.cards = newCard;
  }

  public setIsLoading(state: boolean) {
    this.isLoading = state;
  }

  public setOriginSizeImage(img: { id?: number, url?: string }) {
    this.originSizeImage = img;
  }

  public setComments(comments: { id: number, text: string, date: string }[]) {
    this.comments = comments;
  }
}

export default new PhotoGrid();
