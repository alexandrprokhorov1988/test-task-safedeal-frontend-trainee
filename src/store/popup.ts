import {makeAutoObservable} from "mobx";

class Popup {
  public isLoadingComment = false;

  constructor() {
    makeAutoObservable(this);
  }

  setIsLoadingComment(state: boolean) {
    this.isLoadingComment = state;
  }
}

export default new Popup();
