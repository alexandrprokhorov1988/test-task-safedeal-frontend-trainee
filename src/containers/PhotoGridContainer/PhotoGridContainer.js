import React from 'react';
import {connect} from 'react-redux'
import PhotoGrid from '../../components/PhotoGrid/PhotoGrid';
import mainApi from "../../utils/mainApi";
import {dateParseFromTimestampToString} from "../../utils/helpers";
import {setCards, setComments, setIsLoading, setOriginSizeImage} from "../../redux/actions/photoGridActions";
import Preloader from '../../components/Preloader/Preloader';

function PhotoGridContainer(
  {
    photoGrid,
    setIsLoading,
    setCards,
    setOriginSizeImage,
    setComments,
    onOpen
  }) {

  function handleGetOriginalSizeImage(imageId) {
    if (imageId === photoGrid.currentOriginSizeImage.id) {
      onOpen();
    } else {
      return mainApi.getOriginalSizeImage(imageId)
        .then((res) => {
          setOriginSizeImage({
            id: res.id,
            url: res.url
          });
          const newComments = res.comments.map((comment) => ({
            text: comment.text,
            date: dateParseFromTimestampToString(comment.date),
            id: comment.id
          }));
          setComments(newComments);
          onOpen();
        })
        .catch((err) => console.log(err));
    }
    return true;
  }

  React.useEffect(() => {
    setIsLoading(true);
    mainApi.getInitialImages()
      .then((data) => setCards(data))
      .catch((err) => console.log(err))
      .finally(() => setIsLoading(false));
  }, [setCards, setIsLoading]);

  return (
    <>
      {photoGrid.isLoading ? <Preloader/> :
        <PhotoGrid
          cards={photoGrid.cards}
          onImageClick={handleGetOriginalSizeImage}
        />
      } </>
  );
}

const mapStateToProps = (store) => ({
  photoGrid: store.photoGrid
});

const mapDispatchToProps = (dispatch) => ({
  setIsLoading: (isLoadingBool) => dispatch(setIsLoading(isLoadingBool)),
  setCards: (cardsArr) => dispatch(setCards(cardsArr)),
  setOriginSizeImage: (imgObj) => dispatch(setOriginSizeImage(imgObj)),
  setComments: (commentsArr) => dispatch(setComments(commentsArr))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PhotoGridContainer);
