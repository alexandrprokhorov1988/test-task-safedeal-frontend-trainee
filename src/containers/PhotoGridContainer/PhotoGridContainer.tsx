import React from 'react';
import {connect} from 'react-redux'
import PhotoGrid from '../../components/PhotoGrid/PhotoGrid';
import mainApi from "../../utils/mainApi";
import {dateParseFromTimestampToString} from "../../utils/helpers";
import {setCards, setComments, setIsLoading, setOriginSizeImage} from "../../redux/actions/photoGridActions";
import Preloader from '../../components/Preloader/Preloader';
import {useAppDispatch} from "../../hooks/useAppDispatch";
import {useAppSelector} from "../../hooks/useAppSelector";

interface IPhotoGridContainer {
  onOpen: () => void,
}

const PhotoGridContainer: React.FC<IPhotoGridContainer> = ({onOpen}) => {

  const photoGrid = useAppSelector(state=> state.photoGrid);
  const {setIsLoading, setOriginSizeImage, setCards, setComments} = useAppDispatch();

  function handleGetOriginalSizeImage(imageId: number) {
    if (imageId === photoGrid.currentOriginSizeImage.id) {
      onOpen();
    } else {
      return mainApi.getOriginalSizeImage(imageId)
        .then((res) => {
          setOriginSizeImage({
            id: res.id,
            url: res.url
          });
          const newComments = res.comments.map((comment: { text: string, date: number, id: number }) => ({
            text: comment.text,
            date: dateParseFromTimestampToString(comment.date),
            id: comment.id
          }));
          setComments(newComments);
          onOpen();
        })
        .catch((err) => console.log(err));
    }
    return true;
  }

  React.useEffect(() => {
    setIsLoading(true);
    mainApi.getInitialImages()
      .then((data) => setCards(data))
      .catch((err) => console.log(err))
      .finally(() => setIsLoading(false));
  }, []);

  return (
    <>
      {photoGrid.isLoading ? <Preloader/> :
        <PhotoGrid
          cards={photoGrid.cards}
          onImageClick={handleGetOriginalSizeImage}
        />
      }
      </>
  );
};

// const mapStateToProps = (store) => ({
//   photoGrid: store.photoGrid
// });
//
// const mapDispatchToProps = (dispatch) => ({
//   setIsLoading: (isLoadingBool) => dispatch(setIsLoading(isLoadingBool)),
//   setCards: (cardsArr) => dispatch(setCards(cardsArr)),
//   setOriginSizeImage: (imgObj) => dispatch(setOriginSizeImage(imgObj)),
//   setComments: (commentsArr) => dispatch(setComments(commentsArr))
// });
//
// export default connect(
//   mapStateToProps,
//   mapDispatchToProps
// )(PhotoGridContainer);

export default PhotoGridContainer;
