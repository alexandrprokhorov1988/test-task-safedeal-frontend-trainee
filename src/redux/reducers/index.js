import {combineReducers} from 'redux';
import {mainReducer} from './main';
import {photoGridReducer} from './photoGrid';
import {popupReducer} from './popup';

export const rootReducer = combineReducers({
  main: mainReducer,
  photoGrid: photoGridReducer,
  popup: popupReducer,
});
