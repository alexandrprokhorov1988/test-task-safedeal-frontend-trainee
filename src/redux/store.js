import {applyMiddleware, createStore} from 'redux'
import {rootReducer} from './reducers';
import {logger} from "./middlewares/logger";

export const store = createStore(rootReducer, applyMiddleware(logger));
