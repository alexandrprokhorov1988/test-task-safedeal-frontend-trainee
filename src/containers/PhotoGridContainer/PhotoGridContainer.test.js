import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import {store} from '../../redux/store';
import PhotoGridContainer from './PhotoGridContainer';

describe('PhotoGridContainer container', () => {
  it('PhotoGridContainer container', () => {
    const renderResult = renderer
      .create(
        <Provider store={store}>
          <PhotoGridContainer/>
        </Provider>
      )
      .toJSON();
    expect(renderResult).toMatchSnapshot();
  });


});
