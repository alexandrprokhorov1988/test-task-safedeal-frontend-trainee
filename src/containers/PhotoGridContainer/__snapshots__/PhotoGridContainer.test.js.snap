// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoGridContainer container PhotoGridContainer container 1`] = `
Array [
  <section
    className="photo-grid"
  />,
  " ",
]
`;
